### PR TITLE
SONARJAVA-1157 Remove ASTNode from TreeFactory

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/ast/parser/ListTreeImpl.java [hunk 1]
```diff
@@ -60,14 +60,17 @@ public boolean isLeaf() {
 
   @Override
   public void accept(TreeVisitor visitor) {
+    for (T t : list) {
+      ((Tree) t).accept(visitor);
+    }
     // TODO
-    throw new UnsupportedOperationException("On class: " + getClass().getSimpleName());
+//    throw new UnsupportedOperationException("On class: " + getClass().getSimpleName());
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/ListTreeImpl.java]: MAJOR This block of commented-out lines of code should be removed. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 2]
```diff
@@ -395,14 +367,12 @@ public ClassTreeImpl completeClassDeclaration(
...
-      children.add(extendsKeyword);
-      children.add((AstNode) superClass);
-      partial.completeSuperclass(extendsKeyword, superClass);
+      children.add(extendsClause.get().first());
+      children.add((JavaTree) extendsClause.get().second());
+      partial.completeSuperclass((SyntaxToken) extendsClause.get().first(), extendsClause.get().second());
     }
     if (implementsClause.isPresent()) {
-      InternalSyntaxToken implementsKeyword = InternalSyntaxToken.create(implementsClause.get().first());
+      InternalSyntaxToken implementsKeyword = ((InternalSyntaxToken) implementsClause.get().first());
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MAJOR Remove those useless parentheses. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessParenthesesCheck)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 3]
```diff
@@ -1216,12 +1130,9 @@ public TryStatementTreeImpl newTryCatch(Optional<List<CatchTreeImpl>> catches, O
     }
   }
 
-  public CatchTreeImpl newCatchClause(AstNode catchTokenAstNode, AstNode openParenTokenAstNode, VariableTreeImpl parameter, AstNode closeParenTokenAstNode, BlockTreeImpl block) {
-    InternalSyntaxToken catchToken = InternalSyntaxToken.create(catchTokenAstNode);
-    InternalSyntaxToken openParenToken = InternalSyntaxToken.create(openParenTokenAstNode);
-    InternalSyntaxToken closeParenToken = InternalSyntaxToken.create(closeParenTokenAstNode);
-
-    return new CatchTreeImpl(catchToken, openParenToken, parameter, closeParenToken, block);
+  public CatchTreeImpl newCatchClause(JavaTree catchToken, InternalSyntaxToken openParenToken, VariableTreeImpl parameter, InternalSyntaxToken closeParenToken, BlockTreeImpl block) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MINOR Split this 182 characters long line (which is greater than 180 authorized). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 4]
```diff
@@ -246,10 +230,7 @@ public TypeTree newType(TypeTree basicOrClassType,
     }
   }
 
-  public TypeArgumentListTreeImpl newTypeArgumentList(AstNode openBracketTokenAstNode, Tree typeArgument, Optional<List<AstNode>> rests, AstNode closeBracketTokenAstNode) {
-    InternalSyntaxToken openBracketToken = InternalSyntaxToken.create(openBracketTokenAstNode);
-    InternalSyntaxToken closeBracketToken = InternalSyntaxToken.create(closeBracketTokenAstNode);
-
+  public TypeArgumentListTreeImpl newTypeArgumentList(InternalSyntaxToken openBracketToken, Tree typeArgument, Optional<List<AstNode>> rests, InternalSyntaxToken closeBracketToken) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MINOR Split this 182 characters long line (which is greater than 180 authorized). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 5]
```diff
@@ -780,16 +744,12 @@ public AstNode completeAnnotationTypeMember(ModifiersTreeImpl modifiers, AstNode
     return partial;
   }
 
-  public AstNode completeAnnotationMethod(TypeTree type, AstNode identifierAstNode, MethodTreeImpl partial, AstNode semiTokenAstNode) {
-    partial.complete(type, new IdentifierTreeImpl(InternalSyntaxToken.create(identifierAstNode)), InternalSyntaxToken.create(semiTokenAstNode));
-
+  public AstNode completeAnnotationMethod(TypeTree type, JavaTree identifierToken, MethodTreeImpl partial, InternalSyntaxToken semiToken) {
+    partial.complete(type, new IdentifierTreeImpl(((InternalSyntaxToken) identifierToken)), semiToken);
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MAJOR Remove those useless parentheses. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessParenthesesCheck)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 6]
```diff
@@ -414,73 +384,71 @@ public ClassTreeImpl completeClassDeclaration(
...
     if (enumConstants.isPresent()) {
       for (EnumConstantTreeImpl enumConstant : enumConstants.get()) {
         members.add(enumConstant);
       }
     }
-    if (semicolonTokenAstNode.isPresent()) {
+    if (semicolonToken.isPresent()) {
       // TODO This is a hack
-      members.add(semicolonTokenAstNode.get());
+//      members.add(semicolonToken.get());
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MAJOR This block of commented-out lines of code should be removed. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 7]
```diff
@@ -1042,40 +983,31 @@ public VariableTreeImpl completeVariableDeclarator(AstNode identifierAstNode, Op
...
   }
 
-  public AssertStatementTreeImpl newAssertStatement(AstNode colonToken, ExpressionTree expression) {
-    return new AssertStatementTreeImpl(InternalSyntaxToken.create(colonToken), expression);
+  public AssertStatementTreeImpl newAssertStatement(InternalSyntaxToken colonToken, ExpressionTree expression) {
+    return new AssertStatementTreeImpl(colonToken, expression);
   }
 
-  public IfStatementTreeImpl completeIf(AstNode ifToken, AstNode openParen, ExpressionTree condition, AstNode closeParen, StatementTree statement,
+  public IfStatementTreeImpl completeIf(JavaTree ifToken, InternalSyntaxToken openParenToken, ExpressionTree condition, InternalSyntaxToken closeParenToken, StatementTree statement,
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MINOR Split this 181 characters long line (which is greater than 180 authorized). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 8]
```diff
@@ -169,7 +162,7 @@ public CompilationUnitTreeImpl newCompilationUnit(
       }
     }
 
-    InternalSyntaxToken eofToken = InternalSyntaxToken.create(eof);
+    InternalSyntaxToken eofToken = ((InternalSyntaxToken) eof);
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MAJOR Remove those useless parentheses. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessParenthesesCheck)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 9]
```diff
@@ -1605,50 +1492,44 @@ public ExpressionTree binaryExpression1(ExpressionTree expression, Optional<List
...
+  public TypeCastExpressionTreeImpl newBasicTypeCastExpression(PrimitiveTypeTreeImpl basicType, InternalSyntaxToken closeParenToken, ExpressionTree expression) {
     return new TypeCastExpressionTreeImpl(basicType, closeParenToken, expression);
   }
 
-  public TypeCastExpressionTreeImpl newClassCastExpression(TypeTree type, Optional<List<Tuple<InternalSyntaxToken, Tree>>> classTypes, AstNode closeParenTokenAstNode,
-    ExpressionTree expression) {
-    InternalSyntaxToken closeParenToken = InternalSyntaxToken.create(closeParenTokenAstNode);
 
-    ImmutableList.Builder<Tree> boundsBuilder = ImmutableList.<Tree>builder();
+  public TypeCastExpressionTreeImpl newClassCastExpression(TypeTree type, Optional<List<Tuple<InternalSyntaxToken, Tree>>> classTypes, InternalSyntaxToken closeParenToken, ExpressionTree expression) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MINOR Split this 200 characters long line (which is greater than 180 authorized). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 10]
```diff
@@ -1465,10 +1356,7 @@ public ExpressionTree completeTernaryExpression(ExpressionTree expression, Optio
       expression;
   }
 
-  public ConditionalExpressionTreeImpl newTernaryExpression(AstNode queryTokenAstNode, ExpressionTree trueExpression, AstNode colonTokenAstNode, ExpressionTree falseExpression) {
-    InternalSyntaxToken queryToken = InternalSyntaxToken.create(queryTokenAstNode);
-    InternalSyntaxToken colonToken = InternalSyntaxToken.create(colonTokenAstNode);
-
+  public ConditionalExpressionTreeImpl newTernaryExpression(InternalSyntaxToken queryToken, ExpressionTree trueExpression, InternalSyntaxToken colonToken, ExpressionTree falseExpression) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MINOR Split this 188 characters long line (which is greater than 180 authorized). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 11]
```diff
@@ -915,9 +862,7 @@ public ArgumentListTreeImpl newSingleElementAnnotation(AstNode openParenTokenAst
 
   // Formal parameters
 
-  public FormalParametersListTreeImpl completeParenFormalParameters(AstNode openParenTokenAstNode, Optional<FormalParametersListTreeImpl> partial, AstNode closeParenTokenAstNode) {
-    InternalSyntaxToken openParenToken = InternalSyntaxToken.create(openParenTokenAstNode);
-    InternalSyntaxToken closeParenToken = InternalSyntaxToken.create(closeParenTokenAstNode);
+  public FormalParametersListTreeImpl completeParenFormalParameters(InternalSyntaxToken openParenToken, Optional<FormalParametersListTreeImpl> partial, InternalSyntaxToken closeParenToken) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MINOR Split this 190 characters long line (which is greater than 180 authorized). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 12]
```diff
@@ -961,8 +904,8 @@ public FormalParametersListTreeImpl newVariableArgumentFormalParameter(Optional<
       variable);
   }
 
-  public VariableTreeImpl newVariableDeclaratorId(AstNode identifierAstNode, Optional<List<Tuple<Optional<List<AnnotationTreeImpl>>, Tuple<AstNode, AstNode>>>> dims) {
-    IdentifierTreeImpl identifier = new IdentifierTreeImpl(InternalSyntaxToken.create(identifierAstNode));
+  public VariableTreeImpl newVariableDeclaratorId(JavaTree identifierToken, Optional<List<Tuple<Optional<List<AnnotationTreeImpl>>, Tuple<InternalSyntaxToken, InternalSyntaxToken>>>> dims) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MINOR Split this 190 characters long line (which is greater than 180 authorized). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 13]
```diff
@@ -2401,30 +2245,30 @@ private static TypeTree applyDim(TypeTree expression, @Nullable ArrayTypeTreeImp
   }
 
   @CheckForNull
-  private static ArrayTypeTreeImpl newArrayTypeTreeWithAnnotations(Optional<List<Tuple<Optional<List<AnnotationTreeImpl>>, Tuple<AstNode, AstNode>>>> dims) {
+  private static ArrayTypeTreeImpl newArrayTypeTreeWithAnnotations(Optional<List<Tuple<Optional<List<AnnotationTreeImpl>>, Tuple<InternalSyntaxToken, InternalSyntaxToken>>>> dims) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MINOR Split this 181 characters long line (which is greater than 180 authorized). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 14]
```diff
@@ -1170,41 +1097,28 @@ public StatementExpressionListTreeImpl newStatementExpressions(ExpressionTree ex
...
-    return new ForEachStatementImpl(forKeyword, openParenToken, variable, colonToken, expression, closeParenToken, statement);
+    return new ForEachStatementImpl((InternalSyntaxToken) forTokenAstNode, openParenToken, variable, colonToken, expression, closeParenToken, statement);
   }
 
-  public WhileStatementTreeImpl whileStatement(AstNode whileToken, AstNode openParen, ExpressionTree expression, AstNode closeParen, StatementTree statement) {
-    InternalSyntaxToken whileKeyword = InternalSyntaxToken.create(whileToken);
-    InternalSyntaxToken openParenToken = InternalSyntaxToken.create(openParen);
-    InternalSyntaxToken closeParenToken = InternalSyntaxToken.create(closeParen);
-    return new WhileStatementTreeImpl(whileKeyword, openParenToken, expression, closeParenToken, statement);
+  public WhileStatementTreeImpl whileStatement(JavaTree whileToken, InternalSyntaxToken openParen, ExpressionTree expression, InternalSyntaxToken closeParen, StatementTree statement) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MINOR Split this 184 characters long line (which is greater than 180 authorized). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 15]
```diff
@@ -780,16 +744,12 @@ public AstNode completeAnnotationTypeMember(ModifiersTreeImpl modifiers, AstNode
...
+  public AstNode completeAnnotationMethod(TypeTree type, JavaTree identifierToken, MethodTreeImpl partial, InternalSyntaxToken semiToken) {
+    partial.complete(type, new IdentifierTreeImpl(((InternalSyntaxToken) identifierToken)), semiToken);
     return partial;
   }
 
-  public MethodTreeImpl newAnnotationTypeMethod(AstNode openParenTokenAstNode, AstNode closeParenTokenAstNode, Optional<Tuple<InternalSyntaxToken, ExpressionTree>> defaultValue) {
-    InternalSyntaxToken openParenToken = InternalSyntaxToken.create(openParenTokenAstNode);
-    InternalSyntaxToken closeParenToken = InternalSyntaxToken.create(closeParenTokenAstNode);
-
+  public MethodTreeImpl newAnnotationTypeMethod(InternalSyntaxToken openParenToken, InternalSyntaxToken closeParenToken, Optional<Tuple<InternalSyntaxToken, ExpressionTree>> defaultValue) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MINOR Split this 189 characters long line (which is greater than 180 authorized). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 16]
```diff
@@ -950,9 +895,7 @@ public FormalParametersListTreeImpl prependNewFormalParameter(VariableTreeImpl v
     }
   }
 
-  public FormalParametersListTreeImpl newVariableArgumentFormalParameter(Optional<List<AnnotationTreeImpl>> annotations, AstNode ellipsisTokenAstNode, VariableTreeImpl variable) {
-    InternalSyntaxToken ellipsisToken = InternalSyntaxToken.create(ellipsisTokenAstNode);
-
+  public FormalParametersListTreeImpl newVariableArgumentFormalParameter(Optional<List<AnnotationTreeImpl>> annotations, InternalSyntaxToken ellipsisToken, VariableTreeImpl variable) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MINOR Split this 184 characters long line (which is greater than 180 authorized). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 17]
```diff
@@ -118,16 +117,15 @@ public ModifiersTreeImpl modifiers(Optional<List<ModifierTree>> modifierNodes) {
     return new ModifiersTreeImpl(modifierNodes.get());
   }
 
-  public ModifierKeywordTreeImpl modifierKeyword(AstNode astNode) {
-    JavaKeyword keyword = (JavaKeyword) astNode.getType();
-    return new ModifierKeywordTreeImpl(kindMaps.getModifier(keyword), astNode);
+  public ModifierKeywordTreeImpl modifierKeyword(JavaTree javaTree) {
+    JavaKeyword keyword = (JavaKeyword) javaTree.getType();
+    return new ModifierKeywordTreeImpl(kindMaps.getModifier(keyword), ((InternalSyntaxToken) javaTree));
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MAJOR Remove those useless parentheses. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessParenthesesCheck)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 18]
```diff
@@ -1022,9 +963,9 @@ public VariableDeclaratorListTreeImpl newVariableDeclarators(VariableTreeImpl va
     return new VariableDeclaratorListTreeImpl(variables.build(), children);
   }
 
-  public VariableTreeImpl completeVariableDeclarator(AstNode identifierAstNode, Optional<List<Tuple<Optional<List<AnnotationTreeImpl>>, Tuple<AstNode, AstNode>>>> dimensions,
+  public VariableTreeImpl completeVariableDeclarator(JavaTree identifierToken, Optional<List<Tuple<Optional<List<AnnotationTreeImpl>>, Tuple<InternalSyntaxToken, InternalSyntaxToken>>>> dimensions,
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MINOR Split this 197 characters long line (which is greater than 180 authorized). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

File: java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java [hunk 19]
```diff
@@ -1325,72 +1229,59 @@ public CaseGroupTreeImpl switchGroup(List<CaseLabelTreeImpl> labels, BlockStatem
...
+  public CaseLabelTreeImpl newDefaultSwitchLabel(JavaTree defaultToken, InternalSyntaxToken colonToken) {
+    InternalSyntaxToken defaultSyntaxToken = (InternalSyntaxToken) defaultToken;
+    return new CaseLabelTreeImpl(defaultSyntaxToken, null, colonToken);
   }
 
-  public SynchronizedStatementTreeImpl synchronizedStatement(AstNode synchronizedToken, AstNode openParen, ExpressionTree expression, AstNode closeParen, BlockTreeImpl block) {
-    InternalSyntaxToken synchronizedKeyword = InternalSyntaxToken.create(synchronizedToken);
-    InternalSyntaxToken openParenToken = InternalSyntaxToken.create(openParen);
-    InternalSyntaxToken closeParenToken = InternalSyntaxToken.create(closeParen);
+  public SynchronizedStatementTreeImpl synchronizedStatement(JavaTree synchronizedToken, InternalSyntaxToken openParenToken, ExpressionTree expression, InternalSyntaxToken closeParenToken, BlockTreeImpl block) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/parser/TreeFactory.java]: MINOR Split this 211 characters long line (which is greater than 180 authorized). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

**Comment by SonarTech** [comment]
SonarQube analysis reported 20 issues:
* MAJOR 7 major
* MINOR 13 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Private method 'getLastTokenAstNode' is never used.](https://github.com/SonarSource/sonar-java/blob/20311dfd26b6db6e1a26c5fd192efc4da35f7f9e/java-squid/src/main/java/org/sonar/java/model/JavaTree.java#L583) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

**Comment by m-g-sonar** [comment]
LGTM!

**Pull request merged**